docs(examples): symfony php guide (#7171)
zitadel/zitadel: docs/docs/examples/login/symfony.md

---
title: ZITADEL with Symfony PHP
sidebar_label: Symfony
---

This integration guide demonstrates the recommended way to incorporate ZITADEL into your Symfony PHP application. 
It explains how to enable user login in your application and how to fetch data from the user info endpoint.

By the end of this guide, your application will have login functionality with basic role mapping, access the current user's profile and a user list accessible by admins.

> > This documentation references our [example](https://github.com/zitadel/example-symfony-oidc) on GitHub.

Reviewer: ```suggestion
:::info
This documentation references our [example](https://github.com/zitadel/example-symfony-oidc) on GitHub.
:::
```

## ZITADEL setup

Before we can start building our application, we have to do a few configuration steps in ZITADEL Console.

### Project roles

The Example expects [user roles](guides/integrate/retrieve-user-roles) to be returned after login.
Symfony uses `ROLE_USER` format.
The application will take care of upper-casing and prefixing for us.

> Inside ZITADEL you can user regular lower-case role names without prefix if you prefer.

Reviewer: ```suggestion
Inside ZITADEL, you can use regular lower-case role names without prefixes, if you prefer.
```

> Symfony automatically assigns `ROLE_USER` to any authenticated user.

In your project settings make sure the "Assert Roles On Authentication" is enabled.

![Project settings in console](/img/symfony/project-settings.png)

In the project Role tab, add 2 special roles:

 - `admin`: Assigned to users that need access to the user list.
 - `foo`: Random role for display purposes

A `user` role is not required. This role is assumed by default for any authenticated user in Symfony.

![Project roles in console](/img/symfony/project-roles.png)

> Finally we can assign the roles to users in the project's authorizations tab.

Reviewer: ```suggestion
Finally, we can assign the roles to users in the project's authorizations tab.
```

Select Web application type and continue.
We use [Authorization Code](/apis/openidoauth/grant-types#authorization-code)for our Symfony application.

![Create app in console](/img/symfony/app-create.png)

Select `CODE` in the next step. This makes sure you still get a secret. Note that the secret never gets exposed on the browser and is therefore kept in a confidential environment. Safe the generated 

![Configure app authentication method in console](/img/symfony/app-auth-method.png)

With the Redirect URIs field, you tell ZITADEL where it is allowed to redirect users to after authentication. For development, you can set dev mode to `true` to enable insecure HTTP and redirect to a `localhost` URI.

For the example application we are writing use:

- `http://localhost:8000/login_check` as Redirect URI
- `http://localhost:8000/logout` as post-logout URI.

![Configure app redirects console](/img/symfony/app-redirects.png)

After the final step you are presented with a client ID and secret.
Copy and paste them to a safe location for later use by the application.
The secret will not be displayed again, but you can regenerate one if you loose it.

> ## Symfony setup

Reviewer: ```suggestion
## Setup new Symfony application
```

Now that you have configured your web application on the ZITADEL side, you can proceed with the integration of your Symfony client.
The example is build on a [generated Symfony web app](https://symfony.com/doc/current/setup.html#creating-symfony-applications), using the following command:

> > Skip this step if you are connecting ZITADEL to an existing application.

Reviewer: ```suggestion
:::info
Skip this step if you are connecting ZITADEL to an existing application.
:::
```

```bash
symfony new my_project_directory --version="7.0.*" --webapp
cd my_project_directory
```

> > The remainder of this guide assumes a Symfony project which already includes all web app bundles, such as security, routing and ORM.

Reviewer: ```suggestion
:::info
The remainder of this guide assumes a Symfony project which already includes all web app bundles, such as security, routing and ORM.
```

> > If you are using this guide against an existing project you must make sure the required bundles are installed using the `composer require` command.

Reviewer: ```suggestion
If you are using this guide against an existing project you must make sure the required bundles are installed using the `composer require` command.
:::
```

### Install Symfony dependencies

To connect with ZITADEL through OpenID connect, you need to install the [Symfony OIDC bundle](https://github.com/Drenso/symfony-oidc). Run the following command:

```bash
composer require drenso/symfony-oidc-bundle
```

## Define the Symfony app

### Create a User class

> First we need to create a User class for the database, so we can persist User info between requests. In this case you don't need password authentication.

Reviewer: ```suggestion
First, we need to create a User class for the database, so we can persist user info between requests. In this case you don't need password authentication.
```

Email addresses are not unique for ZITADEL users. There can be multiple user accounts with the same email address.
See [User Constraints](https://zitadel.com/docs/concepts/structure/users#constraints) for more details.
We will use the User Info `sub` claim as unique "display" name for the user. `sub` equals the unique User ID from ZITADEL.
This creates a User Repository and Entity that implements the `UserInterface`:

> > You can skip this step if you already have an existing User object in your project.

Reviewer: ```suggestion
:::info
You can skip this step, if you already have an existing User object in your project.
:::
```

### Controllers and templates

We need to create couple of Controllers and templates to define the app.

#### Index

The index controller serves a public page on the `/` route and provides some basic links to the authenticated sections of the app.

```php reference
https://github.com/zitadel/example-symfony-oidc/blob/main/src/Controller/IndexController.php
```

The index template:

```twig reference
https://github.com/zitadel/example-symfony-oidc/blob/main/templates/index.html.twig
```

#### Login

The login controller initiates the OIDC login flow by creating a Auth request and redirecting the user to ZITADEL.

```php reference
https://github.com/zitadel/example-symfony-oidc/blob/main/src/Controller/LoginController.php
```

#### Profile

The profile controller displays User Info of the currently authenticated user.
Any authenticated user will have access to this page.

```php reference
https://github.com/zitadel/example-symfony-oidc/blob/main/src/Controller/ProfileController.php
```

The profile template maps the User Entity to a HTML page.

```twig reference
https://github.com/zitadel/example-symfony-oidc/blob/main/templates/profile.html.twig
```

#### User list

> The user list controller displays all users from the database that were created during OIDC login.

Reviewer: ```suggestion
The user list controller displays all users from the database, that were created during OIDC login.
```

Only users with an admin role will have access to this page.

```php reference
https://github.com/zitadel/example-symfony-oidc/blob/main/src/Controller/UserListController.php
```

```twig reference
https://github.com/zitadel/example-symfony-oidc/blob/main/templates/user_list.html.twig
```

## Configure and run the application

> > Never store and commit secrets in a `.env` file. Use a `env.local` file instead and make sure the file is in `.gitignore`.

Reviewer: ``` suggestion
:::warning
Never store and commit secrets in a `.env` file. Use a `env.local` file instead and make sure the file is in `.gitignore`.
:::
```
